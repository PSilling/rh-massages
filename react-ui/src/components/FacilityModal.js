// react imports
import React, { Component } from 'react';


// component imports
import AddButton from './AddButton';
import ModalActions from './ModalActions';

// module imports
import {ModalContainer, ModalDialog} from 'react-modal-dialog';

// util imports
import _t from '../utils/Translations.js';
import Util from '../utils/Util.js';

class FacilityModal extends Component {

  state = {name: ""}

  /**
   * Sets default input values on props change.
   */
  componentWillReceiveProps(nextProps) {
    if(this.props === nextProps) return;

    this.setState({
      name: (nextProps.facility === -1) ? "" : nextProps.facility.name
    });
  }

  changeName = (event) => {
    this.setState({name: event.target.value});
  }

  /**
   * Handles the post request.
   */
  addFacility = () => {
    Util.post(Util.FACILITIES_URL, {
      name: this.state.name
    }, () => {
      this.props.onToggle();
      this.props.getCallback();
    });
  }

  /**
   * Handles the put request.
   */
  editFacility = () => {
    Util.put(Util.FACILITIES_URL + this.props.facility.id, {
      name: this.state.name
    }, () => {
      this.props.onToggle();
      this.props.getCallback();
    });
  }

  render() {
    return(
      <div className='pull-right'>
        <AddButton onAdd={this.props.onToggle} />

        {this.props.active ?
          <ModalContainer onClose={this.props.onToggle}>
            <ModalDialog onClose={this.props.onToggle} width="50%">
              <h2>
                {this.props.facility === -1 ?
                  _t.translate('New Facility') : _t.translate('Edit Facility')
                }
              </h2>
              <hr />
              <div className="form-group">
                <label>{ _t.translate('Name') }</label>
                <input value={this.state.name} onChange={this.changeName}
                  className="form-control" />
              </div>
              {this.props.facility === -1 ?
                <ModalActions
                  primaryLabel={ _t.translate('Add') }
                  onProceed={this.addFacility}
                  onClose={this.props.onToggle}
                /> :
                <ModalActions
                  primaryLabel={ _t.translate('Edit') }
                  onProceed={this.editFacility}
                  onClose={this.props.onToggle}
                />
              }
            </ModalDialog>
          </ModalContainer> : ''
        }
      </div>
    )
  }
}

export default FacilityModal
