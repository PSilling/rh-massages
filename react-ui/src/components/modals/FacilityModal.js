// react imports
import React, { Component } from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";

// module imports
import { ModalContainer, ModalDialog } from "react-modal-dialog";

// component imports
import AddButton from "../iconbuttons/AddButton";
import ModalActions from "../buttons/ModalActions";

// util imports
import _t from "../../util/Translations";
import Util from "../../util/Util";

/**
 * Input Modal for Facility management. Allows the change of Facility name.
 * Based on given values can be used for both creating and editing of Facilities.
 */
class FacilityModal extends Component {
  state = { name: "" };

  componentWillReceiveProps(nextProps) {
    if (this.props === nextProps) return;

    this.setState({
      name: nextProps.facility === null ? "" : nextProps.facility.name
    });
  }

  changeName = event => {
    this.setState({ name: event.target.value });
  };

  addFacility = () => {
    if (Util.isEmpty(this.state.name)) {
      Util.notify("error", "", _t.translate("Name is required!"));
      return;
    }
    Util.post(
      Util.FACILITIES_URL,
      {
        name: this.state.name
      },
      () => {
        this.props.onToggle();
        this.props.getCallback();
      }
    );
  };

  editFacility = () => {
    if (Util.isEmpty(this.state.name)) {
      Util.notify("error", "", _t.translate("Name is required!"));
      return;
    }
    Util.put(
      Util.FACILITIES_URL + this.props.facility.id,
      {
        name: this.state.name
      },
      () => {
        this.props.onToggle();
        this.props.getCallback();
      }
    );
  };

  handleModalKeyPress = event => {
    if (event.charCode === 13 && document.activeElement === ReactDOM.findDOMNode(this.modal)) {
      if (this.props.facility === null) {
        this.addFacility();
      } else {
        this.editFacility();
      }
    }
  };

  handleInputKeyPress = event => {
    if (event.charCode === 13) {
      if (this.props.facility === null) {
        this.addFacility();
      } else {
        this.editFacility();
      }
    }
  };

  renderInsides = () => (
    <div>
      <h2>{this.props.facility === null ? _t.translate("New Facility") : _t.translate("Edit Facility")}</h2>
      <hr />
      <div className="form-group col-md-12">
        <label htmlFor="facilityInput">{_t.translate("Name")}</label>
        <input
          id="facilityInput"
          value={this.state.name}
          onChange={this.changeName}
          className="form-control"
          autoFocus
          onFocus={Util.moveCursorToEnd}
          onKeyPress={this.handleInputKeyPress}
          type="text"
          maxLength="64"
          placeholder={_t.translate("Name")}
        />
      </div>
      {this.props.facility === null ? (
        <ModalActions primaryLabel={_t.translate("Add")} onProceed={this.addFacility} onClose={this.props.onToggle} />
      ) : (
        <ModalActions primaryLabel={_t.translate("Edit")} onProceed={this.editFacility} onClose={this.props.onToggle} />
      )}
    </div>
  );

  renderModal = () => {
    if (this.props.withPortal) {
      return (
        <ModalContainer onClose={this.props.onToggle}>
          <ModalDialog
            onClose={this.props.onToggle}
            width="50%"
            style={{ outline: "none" }}
            tabIndex="-1"
            onKeyPress={this.handleModalKeyPress}
            ref={dialog => {
              this.modalDialog = dialog;
            }}
          >
            {this.renderInsides()}
          </ModalDialog>
        </ModalContainer>
      );
    }
    return this.renderInsides();
  };

  render() {
    return (
      <div className="pull-right">
        <AddButton onAdd={this.props.onToggle} />

        {this.props.active && this.renderModal()}
      </div>
    );
  }
}

FacilityModal.propTypes = {
  /** whether the dialog should be shown */
  active: PropTypes.bool.isRequired,
  /** callback function for Facility list update */
  getCallback: PropTypes.func.isRequired,
  /** function called on modal toggle */
  onToggle: PropTypes.func.isRequired,
  /** Facility to be possibly edited or null when adding */
  facility: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string
  }),
  /** whether ModalContainer should be used; useful for testing to avoid portals */
  withPortal: PropTypes.bool
};

FacilityModal.defaultProps = {
  facility: null,
  withPortal: true
};

export default FacilityModal;
