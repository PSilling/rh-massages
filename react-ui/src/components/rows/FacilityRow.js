// react imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// component imports
import DeleteButton from '../iconbuttons/DeleteButton';
import EditButton from '../iconbuttons/EditButton';

/**
 * Custom table row for a facility.
 */
class FacilityRow extends Component {

  shouldComponentUpdate(nextProps) {
    return (this.props.facility.id !== nextProps.facility.id
            || this.props.facility.name !== nextProps.facility.name);
  }

  render() {
    return (
      <tr>
        <td>{this.props.facility.name}</td>
        <td width="105px">
          <span className="pull-right">
            <span style={{ 'marginRight': '5px' }}>
              <EditButton onEdit={this.props.onEdit} />
            </span>
            <DeleteButton onDelete={this.props.onDelete} />
          </span>
        </td>
      </tr>
    );
  }
}

FacilityRow.propTypes = {
  facility: PropTypes.object.isRequired, // the facility for this row
  onEdit: PropTypes.func.isRequired, // function called on Facility edit
  onDelete: PropTypes.func.isRequired, // function called on Facility delete
}

export default FacilityRow
