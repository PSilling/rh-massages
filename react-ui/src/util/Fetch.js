/**
 * Utility class containing functions for client-server communication using the Fetch API or the WebSocket API.
 */

// module imports
import moment from "moment";

// util imports
import _t from "./Translations";
import Auth from "./Auth";
import Util from "./Util";

// const WebSocket = require('ws');
const Fetch = function Fetch() {};

/**
 * Fetches data from a given endpoint.
 * @param url             defined endpoint
 * @param update          callback function to update the resources
 */
Fetch.get = (url, update = () => {}) => {
  Auth.keycloak
    .updateToken(Util.REFRESH_MIN_TIME)
    .success(() => {
      fetch(url, {
        method: "get",
        headers: {
          Authorization: `Bearer ${Auth.getToken()}`
        }
      })
        .then(response => {
          if (response.ok) {
            return response.json();
          }
          console.error("Could not retrieve any data from server! Response: ", response); /* eslint-disable-line */
          return undefined;
        })
        .catch(error => {
          /* eslint-disable-next-line */
          console.error("An unexpected error occurred during the 'GET' request!", error.toString());
          Util.notify(
            "error",
            error.toString(),
            _t.translate("An unexpected error occurred during the 'GET' request!")
          );
        })
        .then(json => {
          update(json);
        });
    })
    .error(() => {
      console.error("Failed to refresh the Keycloak user token!"); /* eslint-disable-line */
      Auth.keycloak.login();
    });
};

/**
 * Creates a new element at a given endpoint.
 * @param url             defined endpoint
 * @param data            data to send
 * @param update          callback function to update the resources
 * @param notify          false if success notifications should be suppressed
 */
Fetch.post = (url, data, update = () => {}, notify = true) => {
  Auth.keycloak
    .updateToken(Util.REFRESH_MIN_TIME)
    .success(() => {
      fetch(url, {
        method: "post",
        credentials: "same-origin",
        headers: {
          Authorization: `bearer ${Auth.getToken()}`,
          "Content-Type": "application/json"
        },
        body: JSON.stringify(data)
      }).then(response => {
        if (response.ok) {
          if (notify) {
            Util.notify("success", "", _t.translate("Your request has been successful."));
          }
          update();
        } else {
          Util.notify(
            "error",
            _t.translate("Your request has ended unsuccessfully."),
            _t.translate("An error occured!")
          );
          console.error("Server responsed with an error response:", response); /* eslint-disable-line */
        }
      });
    })
    .error(() => {
      console.log("Failed to refresh the Keycloak user token!"); /* eslint-disable-line */
      Auth.keycloak.login();
    });
};

/**
 * Edits an element at a given endpoint.
 * @param url             defined endpoint
 * @param data            data to send
 * @param update          callback function to update the resources
 * @param notify          false if success notifications should be suppressed
 * @param onError         function that replaces (if not null) standard error notification
 */
Fetch.put = (url, data, update = () => {}, notify = true, onError = null) => {
  Auth.keycloak
    .updateToken(Util.REFRESH_MIN_TIME)
    .success(() => {
      fetch(url, {
        method: "put",
        credentials: "same-origin",
        headers: {
          Authorization: `bearer ${Auth.getToken()}`,
          "Content-Type": "application/json"
        },
        body: JSON.stringify(data)
      }).then(response => {
        if (response.ok) {
          if (notify) {
            Util.notify("success", "", _t.translate("Your request has been successful."));
          }
          update();
        } else {
          if (onError == null) {
            Util.notify(
              "error",
              _t.translate("Your request has ended unsuccessfully."),
              _t.translate("An error occured!")
            );
          } else {
            onError();
          }
          console.error("Server responsed with an error response:", response); /* eslint-disable-line */
        }
      });
    })
    .error(() => {
      console.log("Failed to refresh the Keycloak user token!"); /* eslint-disable-line */
      Auth.keycloak.login();
    });
};

/**
 * Deletes an element at a given endpoint.
 * @param url             defined endpoint
 * @param update          callback function to update the resources
 * @param notify          false if success notifications should be suppressed
 */
Fetch.delete = (url, update = () => {}, notify = true) => {
  Auth.keycloak
    .updateToken(Util.REFRESH_MIN_TIME)
    .success(() => {
      fetch(url, {
        method: "delete",
        credentials: "same-origin",
        headers: {
          Authorization: `bearer ${Auth.getToken()}`
        }
      }).then(response => {
        if (response.ok) {
          if (notify) {
            Util.notify("success", "", _t.translate("Your request has been successful."));
          }
          update();
        } else {
          Util.notify(
            "error",
            _t.translate("Your request has ended unsuccessfully."),
            _t.translate("An error occured!")
          );
          console.error("Server responsed with an error response:", response); /* eslint-disable-line */
        }
      });
    })
    .error(() => {
      console.log("Failed to refresh the Keycloak user token!"); /* eslint-disable-line */
      Auth.keycloak.login();
    });
};

/**
 * Creates a new WebSocket connection, subscribing to a given URL address. Drops the action if the active WebSocket
 * is trying to connect and was created before the connection timeout limit expired.
 * @see https://github.com/sockjs/sockjs-client for more information about the library.
 * @param webSocketUrl    end of the URL address to subscribe to
 * @return                the created WebSocket
 */
Fetch.subscribeToUrl = webSocketUrl => {
  if (
    Fetch.WEBSOCKET_START_TIME !== null &&
    Fetch.WEBSOCKET.readyState === WebSocket.CONNECTING &&
    moment.duration(moment().diff(Fetch.WEBSOCKET_START_TIME)).asMilliseconds() < Util.WEBSOCKET_TIMEOUT_LIMIT
  ) {
    return Fetch.WEBSOCKET;
  }

  Fetch.WEBSOCKET_AUTHENTICATED = false;
  Fetch.WEBSOCKET_START_TIME = moment();
  const webSocket = new WebSocket(webSocketUrl);

  webSocket.addEventListener("message", event => {
    if (!Fetch.WEBSOCKET_AUTHENTICATED) {
      Fetch.post(
        Util.WEBSOCKETS_URL,
        event.data,
        () => {
          Fetch.WEBSOCKET_AUTHENTICATED = true;
        },
        false
      );
      return;
    }

    // parse the message; format: OPERATION_STATUS or OPERATION_SUBSCRIPTION_DATA
    const operation = event.data.substring(0, event.data.indexOf("_"));
    const remainder = event.data.substring(event.data.indexOf("_") + 1, event.data.length);
    const subscription = remainder.substring(0, remainder.indexOf("_"));

    // send the message to the subscription matching callback function
    if (subscription !== "") {
      const callback = Fetch.WEBSOCKET_CALLBACKS[subscription.toLowerCase()];
      if (typeof callback === "function") {
        callback(operation, JSON.parse(remainder.substring(remainder.indexOf("_") + 1)));
      }
    }
  });

  webSocket.addEventListener("error", event => {
    console.log("WebSocket error. Message: ", event.data); /* eslint-disable-line */
  });

  return webSocket;
};

/**
 * Tries to send a message using the global WebSocket connection. Waits until the WebSocket is ready
 * if unauthenticated or closed (based on retry count and timeout limitations).
 * @param message     the message to be sent
 */
Fetch.tryWebSocketSend = message => {
  if (!Fetch.WEBSOCKET_AUTHENTICATED) {
    setTimeout(() => Fetch.tryWebSocketSend(message), 50);
    return;
  }

  if (Fetch.WEBSOCKET.readyState === WebSocket.OPEN) {
    Fetch.WEBSOCKET.send(message);
  } else {
    const timeouts = [];

    if (Fetch.WEBSOCKET.readyState === WebSocket.CLOSED) {
      Fetch.WEBSOCKET_AUTHENTICATED = false;
    }

    for (let i = 0; i <= Util.WEBSOCKET_RETRY_COUNT; i++) {
      timeouts[i] = setTimeout(() => {
        if (Fetch.WEBSOCKET.readyState === WebSocket.OPEN) {
          Fetch.WEBSOCKET.send(message);
          for (let j = i + 1; j < timeouts.length; j++) {
            clearTimeout(timeouts[j]);
          }
        } else {
          Fetch.WEBSOCKET = Fetch.subscribeToUrl(Util.WEBSOCKET_PROTOCOL_URL);
        }
      }, Util.WEBSOCKET_TIMEOUT_LIMIT * i * i);
    }
  }
};

/** whether the current WebSocket connection is authenticated */
Fetch.WEBSOCKET_AUTHENTICATED = false;
/** connection start time (as a moment) for the current WebSocket */
Fetch.WEBSOCKET_START_TIME = null;
/** global WebSocket connection */
Fetch.WEBSOCKET = Fetch.subscribeToUrl(Util.WEBSOCKET_PROTOCOL_URL);
/** global WebSocket callback object with subscription info before each callback */
Fetch.WEBSOCKET_CALLBACKS = {
  client: null,
  facility: null,
  massage: null
};
/** WebSocket ADD operation code (for POST) */
Fetch.OPERATION_ADD = "ADD";
/** WebSocket CHANGE operation code (for PUT) */
Fetch.OPERATION_CHANGE = "CHANGE";
/** WebSocket REMOVE operation code (for DELETE) */
Fetch.OPERATION_REMOVE = "REMOVE";

export default Fetch;
