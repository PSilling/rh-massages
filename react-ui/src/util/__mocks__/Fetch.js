const get = jest.fn((url, update) => {
  update([]);
});
const post = jest.fn((url, data, update) => {
  update();
});
const put = jest.fn((url, data, update) => {
  update();
});

module.exports = {
  get,
  post,
  put
};
