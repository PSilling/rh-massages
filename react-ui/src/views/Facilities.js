// react imports
import React, { Component } from 'react';


// component imports
import DeleteButton from '../components/DeleteButton';
import EditButton from '../components/EditButton';
import FacilityModal from '../components/FacilityModal';

// util imports
import _t from '../utils/Translations.js';
import Util from '../utils/Util.js';

var facility1 = {
  name: "Fac",
  massages: [
    {
      id: 1,
      date: new Date(),
      user: {name: "User1"},
      masseuse: {name: "Mas1", id: -1},
      facility: {name: "Fac"}
    }
  ]
}

var facility2 = {
  name: "Fac2",
  massages: [
    {
      id: 2,
      date: new Date(0),
      user: null,
      masseuse: {name: "Mas2", id: 1},
      facility: {name: "Fac2"}
    }
  ]
}

/**
 * Tabbed facilities with their massage lists.
 */

class FacilitiesList extends Component {

  state = {facilities: [facility1, facility2], modalActive: false, editId: -1}

  componentDidMount() {
    //this.getFacilities();
  }

  getFacilities = () => {
    Util.get("/api/facilities", (json) => {
      this.setState({facilities: json});
    });
  }

  deleteFacility = (name) => {
    Util.delete("/api/facilities/" + name, this.getFacilities);
  }

  toggleModal = (id) => {
    this.setState({modalActive: !this.state.modalActive, editId: id});
  }

  render () {
    return (
      <div>
        <h1>
          { _t.translate('Facilities') }
        </h1>
        <hr />
        <table className="table table-hover table-responsive">
          <thead>
            <tr>
              <th>{ _t.translate('Name') }</th>
              <th>
                <FacilityModal
                  active={this.state.modalActive}
                  facility={this.state.editId === -1 ?
                    -1 : this.state.facilities[this.state.editId]}
                  getCallback={() => {this.getFacilities()}}
                  onToggle={() => {this.toggleModal(-1)}}
                />
              </th>
            </tr>
          </thead>
          <tbody>
            {this.state.facilities.map((item, index) => (
              <tr key={index}>
                <td>{item.name}</td>
                <td width="105px">
                  <span className="pull-right">
                    <span style={{ 'marginRight': '5px' }}>
                      <EditButton onEdit={() => this.toggleModal(index)} />
                    </span>
                    <DeleteButton onDelete={() => this.deleteFacility(item.name)} />
                  </span>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

const Facilities = () => (
  <div>
    <FacilitiesList />
  </div>
);

export default Facilities
