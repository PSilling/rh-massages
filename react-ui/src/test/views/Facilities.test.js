// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import { Table } from "reactstrap";
import Facilities from "../../views/Facilities";
import FacilityModal from "../../components/modals/FacilityModal";
import FacilityRow from "../../components/rows/FacilityRow";
import Fetch from "../../util/Fetch";

// test mocks
jest.mock("../../util/Auth");
jest.mock("../../util/Fetch");

afterAll(() => {
  jest.resetAllMocks();
});

test("renders content correctly", () => {
  const testFacilities = [{ id: 1, name: "test" }];
  const wrapper = shallow(<Facilities />);

  expect(Fetch.tryWebSocketSend).toHaveBeenCalledTimes(1);
  expect(Fetch.send).toHaveBeenCalledTimes(1);

  wrapper.instance().setState({ facilities: testFacilities });
  const table = wrapper.find(Table);

  expect(table.length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test("properly changes state variables", () => {
  Fetch.delete = jest.fn((url, update = () => {}) => {
    update();
  });
  const testFacilities = [{ id: 1, name: "test" }];
  const wrapper = shallow(<Facilities />);
  const modal = wrapper.find(FacilityModal);
  let rows = wrapper.find(FacilityRow);

  expect(rows.length).toBe(0);
  wrapper.instance().setState({ loading: true, facilities: testFacilities });
  wrapper.instance().getFacilities();
  expect(wrapper.instance().state.facilities).toEqual([]);
  expect(wrapper.instance().state.loading).toBe(false);

  wrapper.instance().setState({ facilities: testFacilities });
  rows = wrapper.find(FacilityRow);
  expect(rows.length).toBe(1);

  expect(modal.props().active).toBe(false);
  expect(modal.props().facility).toBe(null);
  rows.get(0).props.onEdit();
  expect(wrapper.instance().state.modalActive).toBe(true);
  expect(wrapper.instance().state.editId).toBe(0);
  modal.props().onToggle();
  expect(wrapper.instance().state.modalActive).toBe(false);
  expect(wrapper.instance().state.editId).toBe(-1);
});
