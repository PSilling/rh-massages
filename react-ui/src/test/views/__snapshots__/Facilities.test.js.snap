// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Facilities />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <InfoAlert
          onClose={[Function]}
        >
          On this page you can manage facilities in which massages take place.
        </InfoAlert>,
        <h1>
          Facilities
        </h1>,
        <Table
          hover={true}
          responsive={true}
          responsiveTag="div"
          size="sm"
          striped={true}
          tag="table"
        >
          <thead>
            <tr>
              <th
                scope="col"
              >
                Name
              </th>
              <th
                scope="col"
              >
                <FacilityModal
                  active={false}
                  facility={null}
                  getCallback={[Function]}
                  onToggle={[Function]}
                  withPortal={true}
                />
              </th>
            </tr>
          </thead>
          <tbody>
            <FacilityRow
              facility={
                Object {
                  "id": 1,
                  "name": "test",
                }
              }
              onDelete={[Function]}
              onEdit={[Function]}
            />
          </tbody>
        </Table>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "On this page you can manage facilities in which massages take place.",
          "onClose": [Function],
        },
        "ref": null,
        "rendered": "On this page you can manage facilities in which massages take place.",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            "Facilities",
          ],
        },
        "ref": null,
        "rendered": Array [
          false,
          "Facilities",
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <thead>
              <tr>
                <th
                  scope="col"
                >
                  Name
                </th>
                <th
                  scope="col"
                >
                  <FacilityModal
                    active={false}
                    facility={null}
                    getCallback={[Function]}
                    onToggle={[Function]}
                    withPortal={true}
                  />
                </th>
              </tr>
            </thead>,
            <tbody>
              <FacilityRow
                facility={
                  Object {
                    "id": 1,
                    "name": "test",
                  }
                }
                onDelete={[Function]}
                onEdit={[Function]}
              />
            </tbody>,
          ],
          "hover": true,
          "responsive": true,
          "responsiveTag": "div",
          "size": "sm",
          "striped": true,
          "tag": "table",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tr>
                <th
                  scope="col"
                >
                  Name
                </th>
                <th
                  scope="col"
                >
                  <FacilityModal
                    active={false}
                    facility={null}
                    getCallback={[Function]}
                    onToggle={[Function]}
                    withPortal={true}
                  />
                </th>
              </tr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <th
                    scope="col"
                  >
                    Name
                  </th>,
                  <th
                    scope="col"
                  >
                    <FacilityModal
                      active={false}
                      facility={null}
                      getCallback={[Function]}
                      onToggle={[Function]}
                      withPortal={true}
                    />
                  </th>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Name",
                    "scope": "col",
                  },
                  "ref": null,
                  "rendered": "Name",
                  "type": "th",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <FacilityModal
                      active={false}
                      facility={null}
                      getCallback={[Function]}
                      onToggle={[Function]}
                      withPortal={true}
                    />,
                    "scope": "col",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "active": false,
                      "facility": null,
                      "getCallback": [Function],
                      "onToggle": [Function],
                      "withPortal": true,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "th",
                },
              ],
              "type": "tr",
            },
            "type": "thead",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <FacilityRow
                  facility={
                    Object {
                      "id": 1,
                      "name": "test",
                    }
                  }
                  onDelete={[Function]}
                  onEdit={[Function]}
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "class",
                "props": Object {
                  "facility": Object {
                    "id": 1,
                    "name": "test",
                  },
                  "onDelete": [Function],
                  "onEdit": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "tbody",
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <InfoAlert
            onClose={[Function]}
          >
            On this page you can manage facilities in which massages take place.
          </InfoAlert>,
          <h1>
            Facilities
          </h1>,
          <Table
            hover={true}
            responsive={true}
            responsiveTag="div"
            size="sm"
            striped={true}
            tag="table"
          >
            <thead>
              <tr>
                <th
                  scope="col"
                >
                  Name
                </th>
                <th
                  scope="col"
                >
                  <FacilityModal
                    active={false}
                    facility={null}
                    getCallback={[Function]}
                    onToggle={[Function]}
                    withPortal={true}
                  />
                </th>
              </tr>
            </thead>
            <tbody>
              <FacilityRow
                facility={
                  Object {
                    "id": 1,
                    "name": "test",
                  }
                }
                onDelete={[Function]}
                onEdit={[Function]}
              />
            </tbody>
          </Table>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "On this page you can manage facilities in which massages take place.",
            "onClose": [Function],
          },
          "ref": null,
          "rendered": "On this page you can manage facilities in which massages take place.",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              false,
              "Facilities",
            ],
          },
          "ref": null,
          "rendered": Array [
            false,
            "Facilities",
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <thead>
                <tr>
                  <th
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    scope="col"
                  >
                    <FacilityModal
                      active={false}
                      facility={null}
                      getCallback={[Function]}
                      onToggle={[Function]}
                      withPortal={true}
                    />
                  </th>
                </tr>
              </thead>,
              <tbody>
                <FacilityRow
                  facility={
                    Object {
                      "id": 1,
                      "name": "test",
                    }
                  }
                  onDelete={[Function]}
                  onEdit={[Function]}
                />
              </tbody>,
            ],
            "hover": true,
            "responsive": true,
            "responsiveTag": "div",
            "size": "sm",
            "striped": true,
            "tag": "table",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <tr>
                  <th
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    scope="col"
                  >
                    <FacilityModal
                      active={false}
                      facility={null}
                      getCallback={[Function]}
                      onToggle={[Function]}
                      withPortal={true}
                    />
                  </th>
                </tr>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <th
                      scope="col"
                    >
                      Name
                    </th>,
                    <th
                      scope="col"
                    >
                      <FacilityModal
                        active={false}
                        facility={null}
                        getCallback={[Function]}
                        onToggle={[Function]}
                        withPortal={true}
                      />
                    </th>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Name",
                      "scope": "col",
                    },
                    "ref": null,
                    "rendered": "Name",
                    "type": "th",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <FacilityModal
                        active={false}
                        facility={null}
                        getCallback={[Function]}
                        onToggle={[Function]}
                        withPortal={true}
                      />,
                      "scope": "col",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "active": false,
                        "facility": null,
                        "getCallback": [Function],
                        "onToggle": [Function],
                        "withPortal": true,
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": "th",
                  },
                ],
                "type": "tr",
              },
              "type": "thead",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <FacilityRow
                    facility={
                      Object {
                        "id": 1,
                        "name": "test",
                      }
                    }
                    onDelete={[Function]}
                    onEdit={[Function]}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "class",
                  "props": Object {
                    "facility": Object {
                      "id": 1,
                      "name": "test",
                    },
                    "onDelete": [Function],
                    "onEdit": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "tbody",
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
